### PR TITLE
enh(api): Add API endpoint to list host groups

### PR DESCRIPTION
**Pull request by loiclau**
## Description

Add API endpoint to list host groups

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [ ] 20.10.x
- [x] 21.04.x (master)

<h2> How this pull request can be tested ? </h2>

/centreon/api/beta/configuration/hosts/groups

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Centreon/Domain/HostConfiguration/HostGroupService.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    
+    /**
+     * @inheritDoc
+     */
+    public function findHostGroup(int $hgId): array
+    {
+        try {
+            return $this->hostGroupRepository->findHostGroup($hgId);
+        } catch (\Throwable $ex) {
+            throw new HostGroupException(_('Error while searching for the host group'), 0, $ex);
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/HostGroupService.php:64]: shouldn't you use HostGroupException::searchHostGroupsException() ?

File: doc/API/centreon-api-v2.1.yaml [hunk 2]
```diff
@@ -3163,6 +3188,87 @@ components:
...
+          example: 2
+        comment:
+          type: string
+          nullable: true
+          maxLength: 65535
+          description: "Comments on this host group"
+        is_activated:
+          type: boolean
+          description: "Indicates whether the host group is activated or not"
+    Media.NullableImage:
```
Comment (callapa) [inline doc/API/centreon-api-v2.1.yaml:3252]: Still needed?

File: config/routes/Centreon/host_group.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+centreon_application_find_host_groups:
+  methods: GET
+  path: /configuration/hosts/groups
+  controller: 'Centreon\Application\Controller\HostGroupController::findHostGroups'
+  condition: "request.attributes.get('version.is_beta') == true"
```
Comment (callapa) [inline config/routes/Centreon/host_group.yaml:5]: condition: "request.attributes.get('version') >= 2.1"

File: src/Centreon/Domain/HostConfiguration/Exception/HostGroupException.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+declare(strict_types=1);
+
+namespace Centreon\Domain\HostConfiguration\Exception;
+
+/**
+ * This class is designed to contain all exceptions for the context of the host template.
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Exception/HostGroupException.php:26]: Be careful when copying and pasting the code.

File: src/Centreon/Domain/HostConfiguration/Exception/HostGroupException.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @package Centreon\Domain\HostConfiguration\Exception
+ */
+class HostGroupException extends \Exception
+{
+    public static function searchHostGroupsException(\Exception $ex): self
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Exception/HostGroupException.php:32]: Missing phpdoc

File: src/Centreon/Domain/HostConfiguration/Exception/HostGroupFactoryException.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+declare(strict_types=1);
+
+namespace Centreon\Domain\HostConfiguration\Exception;
+
+/**
+ * This class is designed to provide all exceptions for the HostTemplate entity factory.
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Exception/HostGroupFactoryException.php:26]: Be careful when copying and pasting the code.

File: src/Centreon/Domain/HostConfiguration/Interfaces/HostGroupWriteRepositoryInterface.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * This interface gathers all the writing operations on the repository.
+ *
+ * @package Centreon\Domain\HostConfiguration\Interfaces
+ */
+interface HostGroupWriteRepositoryInterface
+{
+    /**
+     * Add a host
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Interfaces/HostGroupWriteRepositoryInterface.php:36]:
```suggestion
     * Add a host group
```

File: src/Centreon/Domain/HostConfiguration/Model/HostGroup.php [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+
+/**
+ * This class is designed to represent a host group.
+ *
+ * @package Centreon\Domain\HostConfiguration
+ */
+class HostGroup
+{
+    public const MAX_NAME_LENGTH = 200,
+        MAX_ALIAS_LENGTH = 200,
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Model/HostGroup.php:36]: keep indentation

File: src/Centreon/Domain/HostConfiguration/Model/HostGroup.php [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+    public function getId(): ?int
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param int|null $id
+     * @return HostGroup
+     */
+    public function setId(?int $id): HostGroup
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Model/HostGroup.php:127]:
```suggestion
    public function setId(int $id): HostGroup
```

File: src/Centreon/Domain/HostConfiguration/Model/HostGroup.php [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+ */
+class HostGroup
+{
+    public const MAX_NAME_LENGTH = 200,
+        MAX_ALIAS_LENGTH = 200,
+        MAX_NOTES = 255,
+        MAX_NOTES_URL = 255,
+        MAX_ACTION_URL = 255,
+        MAX_GEO_COORDS = 32,
+        MAX_RRD = 99999999999,
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/Model/HostGroup.php:41]: ????

File: src/Centreon/Domain/HostConfiguration/UseCase/V21/FindHostGroups.php [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+declare(strict_types=1);
+
+namespace Centreon\Domain\HostConfiguration\UseCase\V21;
+
+use Centreon\Domain\HostConfiguration\Exception\HostGroupException;
+use Centreon\Domain\HostConfiguration\Interfaces\HostGroupReadRepositoryInterface;
+use Centreon\Domain\HostConfiguration\UseCase\V21\FindHostGroupsResponse;
+
+/**
+ * This class is designed to represent a use case to find all host templates
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/UseCase/V21/FindHostGroups.php:30]: Be careful when copying and pasting the code.

File: src/Centreon/Domain/HostConfiguration/UseCase/V21/FindHostGroupsResponse.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+class FindHostGroupsResponse
+{
+    /**
+     * @var array<int, array<string, mixed>>
+     */
+    private $hostGroups = [];
+
+    /**
+     * @param HostGroup[] $hostGroups
+     * @throws \Exception
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/UseCase/V21/FindHostGroupsResponse.php:42]: ?

File: src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+        if ($statement !== false) {
+            while (($result = $statement->fetch(\PDO::FETCH_ASSOC)) !== false) {
+                $hostGroups[] = HostGroupFactoryRdb::create($result);
+            }
+        }
+        return $hostGroups;
+    }
+
+    /**
+     * Add a host group.
```
Comment (callapa) [inline src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php:191]: Descriptions must be defined in the interface only.

File: src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+use Centreon\Domain\HostConfiguration\Model\HostGroup;
+use Centreon\Domain\Repository\RepositoryException;
+use Centreon\Domain\RequestParameters\RequestParameters;
+use Centreon\Infrastructure\DatabaseConnection;
+use Centreon\Infrastructure\HostConfiguration\Repository\Model\HostGroupFactoryRdb;
+use Centreon\Infrastructure\Repository\AbstractRepositoryDRB;
+use Centreon\Infrastructure\RequestParameters\SqlRequestParametersTranslator;
+
+/**
+ * This class is designed to represent the MariaDb repository to manage host, host group and host template
```
Comment (callapa) [inline src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php:36]: Are you sure ?

File: src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+            : ' GROUP BY hg.hg_id';
+
+        // Sort
+        $sortRequest = $this->sqlRequestTranslator->translateSortParameterToSql();
+        $request .= !is_null($sortRequest)
+            ? $sortRequest
+            : ' ORDER BY hg.hg_id ASC';
+        // Pagination
+        $request .= $this->sqlRequestTranslator->translatePaginationToSql();
+        $statement = $this->db->query($request);
```
Comment (callapa) [inline src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php:174]: Why apply the search parameters if you don't take them into account?

File: src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+     * @throws RepositoryException
+     * @throws \Exception
+     */
+    private function addHostGroupRelation(int $hostId, array $hostGroups): void
+    {
+        if (empty($hostGroups)) {
+            return;
+        }
+
+        foreach ($hostGroups as $order => $hostGroup) {
```
Comment (callapa) [inline src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php:204]: $order is never used.

File: src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+                    'INSERT INTO `:db`.hostgroup_relation
+                    (`host_host_id`, `hostgroup_hg_id`)
+                    VALUES (:host_id, :hg_id)'
+                );
+                $statement = $this->db->prepare($request);
+                $statement->bindValue(':host_id', $hostId, \PDO::PARAM_INT);
+                $statement->bindValue(':hg_id', $hostGroup->getId(), \PDO::PARAM_INT);
+                $statement->execute();
+                if ($statement->rowCount() === 0) {
+                    throw new RepositoryException(sprintf(_('Host Group with id %d not found'), $hostGroup->getId()));
```
Comment (callapa) [inline src/Centreon/Infrastructure/HostConfiguration/Repository/HostGroupRepositoryRDB.php:217]: You forgot to use a transaction to cancel previous insertions.

File: src/Centreon/Infrastructure/HostConfiguration/Repository/Model/HostGroupFactoryRdb.php [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+declare(strict_types=1);
+
+namespace Centreon\Infrastructure\HostConfiguration\Repository\Model;
+
+use Centreon\Domain\HostConfiguration\Model\HostGroup;
+use Centreon\Domain\Media\Model\Image;
+use Centreon\Domain\HostConfiguration\Exception\HostGroupFactoryException;
+
+/**
+ * This class is designed to provide a way to create the HostTemplate entity from the database.
```
Comment (callapa) [inline src/Centreon/Infrastructure/HostConfiguration/Repository/Model/HostGroupFactoryRdb.php:30]: Be careful when copying and pasting the code.

File: tests/php/Centreon/Infrastructure/HostConfiguration/Model/HostGroupFactoryRdbTest.php [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+class HostGroupFactoryRdbTest extends TestCase
+{
+    /**
+     * @var array<string, string|int> $rdbData
+     */
+    private $rdbData;
+
+    protected function setUp(): void
+    {
+        $this->rdbData = [
```
Comment (callapa) [inline tests/php/Centreon/Infrastructure/HostConfiguration/Model/HostGroupFactoryRdbTest.php:41]: rddData or hostGroup ?

File: src/Centreon/Domain/HostConfiguration/HostGroupService.php [hunk 20]
```diff
@@ -0,0 +1,79 @@
+<?php
```
Comment (callapa) [inline src/Centreon/Domain/HostConfiguration/HostGroupService.php:1]: Be careful when copying and pasting the code.
What is the purpose of this class, where do you use it?

**Review by callapa — CHANGES_REQUESTED** [review]
Please to change the date in the copyright also.

File: src/Centreon/Domain/HostConfiguration/Model/HostGroup.php [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+    public function getId(): ?int
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param int $id
+     * @return HostGroup
+     */
+    public function setId(int $id): HostGroup
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/Model/HostGroup.php:137]:
```suggestion
    public function setId(?int $id): HostGroup
```
    /**
     * @var int|null
     */
    private $id;

File: src/Centreon/Domain/HostConfiguration/Model/HostGroup.php [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+    /**
+     * @return int|null
+     */
+    public function getId(): ?int
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param int $id
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/Model/HostGroup.php:134]:
```suggestion
     * @param int|null $id
```

File: src/Centreon/Domain/HostConfiguration/Model/HostGroup.php [hunk 23]
```diff
@@ -0,0 +1,373 @@
...
+    /**
+     * @param string $name
+     * @return HostGroup
+     * @throws \Assert\AssertionFailedException
+     */
+    public function setName(string $name): HostGroup
+    {
+        if ($name !== null) {
+            Assertion::maxLength($name, self::MAX_NAME_LENGTH, 'HostGroup::name');
+        }
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/Model/HostGroup.php:160]: in method definition $name can't be null , I don't really know if you have to remove this condition, or add string|null to property name (and adequate changes in getter/setter), depending of the needed behaviour.

File: src/Centreon/Domain/HostConfiguration/UseCase/V21/HostGroup/FindHostGroups.php [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+ * This class is designed to represent a use case to find all host groups
+ *
+ * @package Centreon\Domain\HostConfiguration\UseCase
+ */
+class FindHostGroups
+{
+    /**
+     * @var HostGroupReadRepositoryInterface
+     */
+    private $configurationReadRepository;
```
Comment (jeremyjaouen) [inline src/Centreon/Domain/HostConfiguration/UseCase/V21/HostGroup/FindHostGroups.php:39]: IMO, i found this naming a little confusing, why not called this  `$hostGroupReadRepository` ?

File: src/Centreon/Infrastructure/HostConfiguration/API/Model/HostGroupV21.php [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * @var int|null
+     */
+    public $id;
+
+    /**
+     * @var string|null
+     */
+    public $name;
```
Comment (jeremyjaouen) [inline src/Centreon/Infrastructure/HostConfiguration/API/Model/HostGroupV21.php:41]: Did the property of the Model should be in adequation with your src/Centreon/Domain/HostConfiguration/Model/HostGroup ? If so those two properties don't match.

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

**Pull request closed without merging**